### ISSUE TITLE
Categorial-like distributions are rejected by CategoricalGibbsMetropolis stepper

### ISSUE THREAD

**Issue by hstrey** [pymc-devs/pymc#1739]
I wrote a more detailed question on StackOverflow [here](http://stackoverflow.com/questions/41899454/what-pymc3-monte-carlo-stepper-can-i-use-for-a-custom-categorial-distribution).

The basic problem is.  I created a custom discrete distribution with a logp method that uses multi-dimensional Categorical distributions.  Unfortunately, when I try to sample this distribution, the CategoricalGibbsMetropolis rejects my custom distribution because its parent distribution is not Categorical.  To solve the problem, I hacked my pymc3 distribution and removed that test and it works with my distribution.

My suggestion for a solution is to either add a flag to the steppers that essentially says: Hey I know its risky, but believe me that I am Categorical, or Binary, or something else.  A more pythonic way would be to use duck typing.  Can I set a class property that will allow the stepper to know whether my distribution is ok for this stepper.  This would also help with the automatic assignment of steppers.

**Comment by hstrey** [comment]
Here is a quick fix that worked for me:
In metropolis.py in the beginning of __init__ I inserted a hasattr(distr, 'k')

     if isinstance(distr, pm.Categorial) or hasattr(distr, 'k'):
        k = draw_values([distr.k])[0]

There is only one discrete distribution that defines k and that is Categorical.  So by testing whether k is an attribute would work.

Similarly, in BinaryGibbsMetropolis one can write something like:
elif (hasattr(distribution, 'k') and (distribution.k ==2):

Should I make the changes and then do a pull request?  I have never done that before, but I can try

**Comment by hstrey** [comment]
After some discussion with Chris, I found a nice solution by subclassing Categorical.

I stand corrected.  I did what you suggested and it seems to work.  Here is the solution:
I define my HMMStates class as a subclass of pm.Categorical
in __init__, instead of calling the super __init__, I call:
super(pm.Categorial, self).__init__(*args, **kwargs)

That makes my custom distribution pass the test for BinaryGibbsMetropolis and CategoricalGibbsMetropolis.

I am closing this issue.
